perf/lua/gh-7089-vclock-copy.lua

```diff
@@ -0,0 +1,203 @@
+            read_only = false,
+            replication = {%s},
+            log = 'log.log',
+        }
+        box.once('bootstrap', function()
+            box.schema.user.grant('guest', 'replication')
+        end)
+
+        repeat
+            fiber.sleep(0.1)
+        until not (#box.info.replication < %d or box.info().status ~= 'running')
+
+        -- This is executed on every instance so that vclock is
+        -- non-empty in each component. This will make the testing
+        -- instance copy a larger portion of data on each write
+        -- and make the performance degradation, if any.
+        box.space._schema:replace({'Something to bump vclock ' .. %d})
+    ]], listen, work_dir, table.concat(master_nodes, ','), nodes, listen)}
+    local res, err = popen.new(cmd)
+
+    if not res then
+        exit(1, 'error running replica: ' .. err)
+    end
+
+    nodes_ph[listen] = res
+end
+
+
+if (nodes ~= nil and nodes < 32 and nodes > 0) then
+    print('# starting ' .. nodes .. ' masters')
+    for listen = 3302, 3301 + nodes do
+        bootstrap_node(listen)
+    end
+else
+    exit(1, 'Incorrect number of nodes: "' .. arg[1] .. '" must be 1..31')
+end
+
+box.cfg{
+    listen = LISTEN_PORT,
+    replication_threads = nodes,
+    -- Disable WAL on a node to notice slightest differences in TX
+    -- thread performance. It's okay to replicate _to_ a node with
+    -- disabled WAL. You only can't replicate _from_ it.
+    wal_mode = wal_mode,
+    replication = master_nodes,
+    read_only = false,
+    log = 'test.log',
+    work_dir = test_dir,
+}
+
+-- Wait for all nodes to connect.
+repeat
+    fiber.sleep(0.1)
+    print('# replication', #box.info.replication,
+          'status ', box.info().status)
+until not (#box.info.replication < nodes or box.info().status ~= 'running')
+
+box.schema.space.create('test', {if_not_exists = true})
+box.space.test:create_index('pk', {if_not_exists = true})
+box.snapshot()
+
+local function replace_func(num_iters)
+    for i = 1, num_iters do
+        box.space.test:replace({i, i})
+    end
+end
+
+local num_replaces = 1e6
+
+local function test(num_fibers)
+    local fibers = {}
+    local num_iters = num_replaces / num_fibers
+    local start_realtime = clock.time()
+    local start_cputime = clock.proc()
+    for _ = 1, num_fibers do
+        local fib = fiber.new(replace_func, num_iters)
+        fib:set_joinable(true)
+        table.insert(fibers, fib)
+    end
+    assert(#fibers == num_fibers, 'Fibers created successfully')
+    for _, fib in pairs(fibers) do
+        fib:join()
+    end
+    local dt_realtime = clock.time() - start_realtime
+    local dt_cputime = clock.proc() - start_cputime
+    return dt_realtime, dt_cputime, num_replaces / dt_realtime
+end
+
+local num_iters = 40
+
+-- Fiber count > 1 makes no sense for `wal_mode = 'none'`. There
+-- are no yields on replace when there are no WAL writes.
+local num_fibers = wal_mode == 'none' and 1 or num_iters
```

Comment:
This was an accident that `num_iters` and `num_fibers` matched. I'd keep `num_fibers` equal to 100 for `wal_mode = 'write'`.

Reply:
Hi, Sergey!
Thanks for the review. Replaced back to 100.